packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerUtils.kt

```diff
@@ -1,92 +1,149 @@
+  } catch (cause: IOException) {
+    throw FailedToCreateFileException(filePath, cause)
   }
 }
 
-fun getType(contentResolver: ContentResolver, uri: Uri): String? {
-  return contentResolver.getType(uri).ifNull {
-    // previous method sometimes returns null
-    getTypeFromFileUrl(uri.toString())
-  }
+internal fun getType(contentResolver: ContentResolver, uri: Uri): String =
+  contentResolver.getType(uri)
+    ?: getTypeFromFileUrl(uri.toString())
+    ?: throw FailedDeducingTypeException()
+
+private fun getTypeFromFileUrl(url: String): String? {
+  val extension = MimeTypeMap.getFileExtensionFromUrl(url)
+  return extension?.let { MimeTypeMap.getSingleton().getMimeTypeFromExtension(it) }
 }
 
-fun contentUriFromFile(file: File, application: Application): Uri {
+/**
+ * Convert this [File] to [Uri] that might be accessed by 3rd party Activities, eg. by camera application
+ */
+internal fun File.toContentUri(context: Context): Uri {
   return try {
-    FileProvider.getUriForFile(application, application.packageName + ".ImagePickerFileProvider", file)
+    FileProvider.getUriForFile(context, context.packageName + ".ImagePickerFileProvider", this)
   } catch (e: Exception) {
-    Uri.fromFile(file)
+    Uri.fromFile(this)
   }
 }
 
-// http://stackoverflow.com/a/38858040/1771921
-fun uriFromFile(file: File): Uri = Uri.fromFile(file)
+internal fun File.toBitmapCompressFormat(): Bitmap.CompressFormat = when {
+  this.extension.contains("png") -> Bitmap.CompressFormat.PNG
+  else -> Bitmap.CompressFormat.JPEG
+}
 
-fun uriFromFilePath(path: String) = uriFromFile(File(path))
+internal fun Bitmap.CompressFormat.toFileExtension(): String {
+  return when(this) {
+    Bitmap.CompressFormat.PNG -> ".png"
+    Bitmap.CompressFormat.JPEG -> ".jpeg"
+    else -> throw RuntimeException("Compress format not supported '${this.name}'")
+  }
+}
 
-fun deduceExtension(type: String): String = when {
-  type.contains("png") -> ".png"
-  type.contains("gif") -> ".gif"
-  type.contains("bmp") -> ".bmp"
-  !type.contains("jpeg") -> {
-    Log.w(ImagePickerConstants.TAG, "Image type not supported. Falling back to JPEG instead.")
-    ".jpg"
+internal fun String.toFileExtension(): String = when {
+  this.contains("png") -> ".png"
+  this.contains("gif") -> ".gif"
+  this.contains("bmp") -> ".bmp"
+  !this.contains("jpeg") -> {
+    Log.w(TAG, "Image type not supported. Falling back to JPEG instead.")
+    ".jpeg"
   }
-  else -> ".jpg"
+  else -> ".jpeg"
 }
 
-class ExifDataHandler(private val uri: Uri) {
-  fun copyExifData(newUri: Uri, contentResolver: ContentResolver) {
-    if (uri == newUri) {
-      return
+internal fun String.toBitmapCompressFormat(): Bitmap.CompressFormat = when {
+  this.contains("png") ||
+  this.contains("gif") ||
+  this.contains("bmp") -> {
+    // The result image won't ever be a GIF of a BMP as the cropper doesn't support it.
+    Bitmap.CompressFormat.PNG
+  }
+  else -> {
+    if (!this.contains("jpeg")) {
```

Comment:
Similar to the comment above. 

I would choose a separate case for `contains("jpeg")` and handle that unknown type in the `else` block.